students/MzKhan/lesson05/mailroom3.py

```diff
@@ -0,0 +1,177 @@
+
+def email(donor_name, amount):
+    """
+    generate the email for the given donor for his/her donation.
+    parm: string
+    parm: float
+    """
+    header = "\nDear {}".format(donor_name)
+    body = "\nThank you so much for your generous donation of $ {:.2f}."
+    print(header, body.format(amount))
+
+def print_report():
+    """Print the donors' report in the formatted tabular form."""
+    title = "{:24} | {:12} | {:10} | {:20}"
+    dashes=67*('-');print(dashes)
+    print(title.format('Donor Name','Total Given','Num Gifts','Average Gift'))
+    strf_format = "{:24} ${:12.2f}   {:^10d}  ${:12.2f}"
+    print(dashes)
+    report_list = calculate_total_gift()
+    for donor in sorted(report_list, key = igetter(1), reverse = True):
+        print(strf_format.format(*donor))
+    print(dashes)
+def calculate_total_gift():
+    """
+    create a list from the global donors' data.
+    return: list
+    """
+    new_list = [ [donor, sum(donors_data[donor][0:]),len(donors_data[donor]),
+               sum(donors_data[donor][0:])/len(donors_data[donor])]
+               for donor in donors_data ]
+    return new_list
+
+def quit():
+    """exit out from the user interaction."""
+    print("Thank you. Have a nice day:")
+    sys.exit()
+
+def send_letter_everyone():
+    """generate the thank you letters for each donor in the list"""
+    date_format = '{:%m-%d-%Y}'.format(datetime.datetime.now())
+    for donor in donors_data:
+        _donor = []
+        donor_donations = donors_data.get(donor)
+        _donor.extend([donor_donations[-1],sum(donor_donations[:]),
+                                             len(donor_donations)])
+        letter=letter_format().format(donor,*_donor)
+        file_name = donor+"_"+date_format+'.txt'
+        write_a_letter(file_name,letter)
+
+def write_a_letter(filename, content):
+    """create a new file in the Letters folder and write to it"""
+    folder = "Letters"
+    if not os.path.exists(folder):
+        os.mkdir(folder)
+    with open(os.path.join(folder,filename),'w+') as out:
+        out.write(content)
+
+def letter_format():
+    """return the letter message"""
+    message = """Dear {:},
+
+    Thank you so much for your kind donation of ${:.2f}. With that you have
+    generously donated a total amount of ${:.2f} in your {} donation(s).
+    We must ensure you that your donations will be put to a very good use.
+
+                                                        Sincerely,
+
+                                                        -Team """
+    return message
+
+
+if __name__ == "__main__":
+
+    #Main program flow control and the interaction with the user.
+    options = {1:thank_you,2:print_report,3:send_letter_everyone,4:quit}
+    while True:
+        try:
+            user_response = prompt_user()
+            options[user_response]()
+        except KeyError as e:
+            print("Invalid Option Selected. Please try again :) ")
+        except ValueError as e:
```

Comment:
you should group KeyError and Value error together - as both are considered invalid options: 'except (KeyError, ValueError)`